### ISSUE TITLE
Bug: [no-unnecessary-condition] False positive with `$.each()`

### ISSUE THREAD

**Issue by marekdedic** [typescript-eslint/typescript-eslint#7210]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.1.6&fileType=.ts&code=GYVwdgxgLglg9mABMMAKAlIg3gKEYgGwFMpEIALIiAayIBMA1AQwJCMQF5kWBnIgbhx4yCHqSLEAtkTBQenRAG0ATABpEAZnUAWALqDhAEgB0RJhVQSi02T3WoA%2Buqs2omDgD5sw-DGCJLKRlSDlDNTCxEH3wyShp6ZlZ2LigAJzZBGMQAX2Fs9CFff1QKKlpGFjYI6IhROGJjAjgAc1QAIgAVdKI29EycnGygA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
function fn() {
  let checkedValue = false;

  const elements = [2, 3, 4];

  $.each(elements, (_, element) => {
    if (element === 3) { 
      checkedValue = true;
    }
  })

  if (checkedValue) {
    console.log("True");
  }
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No issue, since the value is overwritten in the for cycle

### Actual Result

Error: `checkedValue` is always falsy

**Comment by bradzacher** [comment]
This is working as expected.
TS has no way to understand that `$.each` immediately calls the passed callback - so it assumes it doesn't.

[Which means that TS does not widen the type of the variable](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMMAKAlIg3gKEYgGwFMpEIALIiAayIBMA1AQwJCMQF5kWBnIgbhx4yCHqSLEAtkTBQenRAG0ATABpEAZnUAWALqDhAEgB0RJhVQSi02T3WoA+uqs2omDgD5sw-DGCJLKRlSDlDNTCxEH3wyShp6ZlZ2LigAJzZBGMQAX2Fs9CFff1QKKlpGFjYI6IB6GvwAPQB+aIhROGJjAjgAc1QAIgAVdKJ+9EycnGygA), meaning it still thinks the variable is typed as `false`.

The rule is solely powered by types - so because TS tells us the variable has the literal type `false`, we report.

You must annotate the variable to force TS to widen the type.

Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/4624 https://github.com/typescript-eslint/typescript-eslint/issues/4501 https://github.com/typescript-eslint/typescript-eslint/issues/4315 https://github.com/typescript-eslint/typescript-eslint/issues/3732 https://github.com/typescript-eslint/typescript-eslint/issues/3270 https://github.com/typescript-eslint/typescript-eslint/issues/2262 https://github.com/typescript-eslint/typescript-eslint/issues/1984 https://github.com/typescript-eslint/typescript-eslint/issues/989 etc

**Comment by marekdedic** [comment]
Hmm, that's too bad - judging by the number of linked issues, I am not the only one confused by this "gotcha", but reading https://github.com/typescript-eslint/typescript-eslint/issues/4624#issuecomment-1057907305 I understand there isn't really much to do here :(

Maybe reopen this issue as docs improvement request? I don't see this caveat mentioned in the [docs](https://typescript-eslint.io/rules/no-unnecessary-condition) - I can do a PR if you're open to it ;)

**Comment by JoshuaKGoldberg** [comment]
+1, makes sense. Filed #7264 to track it specifically so that this one still shows up for other folks using `$.each`.

**Comment by jfmengels** [comment]
> meaning it still thinks the variable is typed as false.
Is there an issue in TypeScript to improve this behaviour?

**Comment by bradzacher** [comment]
https://github.com/microsoft/TypeScript/issues/9998
This issue and a lot of the related issues that reference it are a good start.

There are ways to solve this in the type system (eg rust has a few function types to help the borrow checker understand the life cycle of a closure) - but they are not small changes in how TS works so I wouldn't hold my breath waiting for this to get solved.

Where possible I would just look to avoid callback style when you need side-effects (eg use a loop instead of `forEach`).

**Comment by Josh-Cena** [comment]
In this particular case, `false as boolean` is sufficient.
